driver-sync/src/test/functional/com/mongodb/internal/connection/OidcAuthenticationProseTests.java

```diff
@@ -0,0 +1,937 @@
+import com.mongodb.MongoSecurityException;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.TestListener;
+import com.mongodb.event.CommandListener;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonArray;
+import org.bson.BsonBoolean;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonString;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.Executable;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.opentest4j.AssertionFailedError;
+import org.opentest4j.MultipleFailuresError;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.mongodb.MongoCredential.ALLOWED_HOSTS_KEY;
+import static com.mongodb.MongoCredential.IdpInfo;
+import static com.mongodb.MongoCredential.OidcRefreshContext;
+import static com.mongodb.MongoCredential.OidcRequestCallback;
+import static com.mongodb.MongoCredential.OidcRequestContext;
+import static com.mongodb.MongoCredential.PROVIDER_NAME_KEY;
+import static com.mongodb.MongoCredential.REFRESH_TOKEN_CALLBACK_KEY;
+import static com.mongodb.MongoCredential.REQUEST_TOKEN_CALLBACK_KEY;
+import static com.mongodb.MongoCredential.createOidcCredential;
+import static com.mongodb.client.TestHelper.setEnvironmentVariable;
+import static java.lang.System.getenv;
+import static java.util.Arrays.asList;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+import static util.ThreadTestHelpers.executeAll;
+
+
+/**
+ * See
+ * <a href="https://github.com/mongodb/specifications/blob/master/source/auth/tests/mongodb-oidc.rst#mongodb-oidc">Prose Tests</a>.
+ */
+public class OidcAuthenticationProseTests {
+
+    public static boolean oidcTestsEnabled() {
+        return Boolean.parseBoolean(getenv().get("OIDC_TESTS_ENABLED"));
+    }
+
+    private static final String AWS_WEB_IDENTITY_TOKEN_FILE = "AWS_WEB_IDENTITY_TOKEN_FILE";
+
+    public static final String TOKEN_DIRECTORY = "/tmp/tokens/"; // TODO-OIDC
```

Comment:
My understanding is that commits to `master` must advance the code from a state that meets certain standard (tests and checks pass on Evergreen, the code passed the review process, there are no TODOs in the new code, sync/async APIs are consistent with each other, etc.) to another such state. Based on the aforementioned understanding and given the following, I think we should change the target branch of this PR from `master` to an OIDC feature branch:

- this PR does not contain the async implementation;
- not all tests added by the PR run on Evergreen.

We can then merge the async PR to that same feature branch, and when the code in the branch meets the the standard for the `master`, approve the PR for merging from the feature branch to `master`.